src/IssueViz/Editor/LocationTagging/LocationTagger.cs

```diff
@@ -0,0 +1,188 @@
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.VisualStudio;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Tagging;
+using SonarLint.VisualStudio.Integration;
+using SonarLint.VisualStudio.IssueVisualization.Models;
+
+namespace SonarLint.VisualStudio.IssueVisualization.Editor.LocationTagging
+{
+    // TODO: track document renames: https://github.com/SonarSource/sonarlint-visualstudio/issues/1662
+
+    // Buffer tagger for Sonar issue locations
+    // * adds location tags for primary and secondary locations for a file
+    // * tracks buffer edits, re-calculate spans, and notifies the issues location store about the changes
+    internal class LocationTagger : ITagger<IIssueLocationTag>, IDisposable
+    {
+        private readonly ITextBuffer buffer;
+        private readonly IIssueLocationStore locationService;
+        private readonly IIssueSpanCalculator spanCalculator;
+        private readonly ILogger logger;
+
+        internal /* for testing */ IList<ITagSpan<IIssueLocationTag>> TagSpans { get; private set; }
+
+        public LocationTagger(ITextBuffer buffer, IIssueLocationStore locationService, IIssueSpanCalculator spanCalculator, ILogger logger)
+        {
+            this.buffer = buffer;
+            this.locationService = locationService;
+            this.spanCalculator = spanCalculator;
+            this.logger = logger;
+
+            FilePath = GetFileName(buffer);
+            UpdateTags();
+
+            locationService.IssuesChanged += OnIssuesChanged;
+            buffer.ChangedLowPriority += SafeOnBufferChange;
+        }
+
+        private void OnIssuesChanged(object sender, IssuesChangedEventArgs e)
+        {
+            if (!e.AnalyzedFiles.Contains(FilePath, StringComparer.OrdinalIgnoreCase))
+            {
+                return; // no changes in this file
+            }
+
+            UpdateTags();
+        }
+
+        private void UpdateTags()
+        {
+            var textSnapshot = buffer.CurrentSnapshot;
+
+            // Get the new locations and calculate the spans if necessary
+            var locations = locationService.GetLocations(FilePath);
+            EnsureSpansExist(locations, textSnapshot);
+
+            TagSpans = locations.Select(CreateTagSpan).ToList();
+
+            // Notify the editor that our set of tags has changed
+            var wholeSpan = new SnapshotSpan(textSnapshot, 0, textSnapshot.Length);
+            TagsChanged?.Invoke(this, new SnapshotSpanEventArgs(wholeSpan));
+        }
+
+        private void EnsureSpansExist(IEnumerable<IAnalysisIssueLocationVisualization> locVizs, ITextSnapshot currentSnapshot)
+        {
+            // All spans for issues in the primary file should have been calculated already (whether primary or secondary).
```

Comment:
Instead of " in the primary file" - maybe "in the current file"?

Also, what would've calculate it? The converter? (perhaps add a `<cref>` to it)

Reply:
I think it's ok to document a general assumption (e.g. most spans will have been set when the analysis was done") as part of explaining why we're calculating some paths here, but I don't think we should have implementation-specific information (particularly since the converter classes are going to be changed).